cluster/src/dunit/scala/io/snappydata/cluster/ConcurrentQueryRoutingDUnitSecureTest.scala

```diff
@@ -0,0 +1,465 @@
+    val serverHostPort = AvailablePortHelper.getRandomAvailableTCPPort
+    vm2.invoke(classOf[ClusterManagerTestBase], "startNetServer", serverHostPort)
+    // scalastyle:off println
+    println(s"network server started at $serverHostPort")
+    // scalastyle:on println
+
+    val tableName = "order_line_row"
+
+    try {
+      createTable2(serverHostPort, jdbcUser3 + "." + tableName, jdbcUser4, jdbcUser4)
+      assert(false) // fail
+    } catch {
+      case x: SQLException if x.getSQLState.equals("42507") ||
+          x.getSQLState.equals("42508") => // ignore
+      case t: Throwable => throw t
+    }
+    createTable2(serverHostPort, tableName, jdbcUser3, jdbcUser3)
+
+    try {
+      insertRows3(20, serverHostPort, jdbcUser3 + "." + tableName, jdbcUser4, jdbcUser4)
+      assert(false) // fail
+    } catch {
+      case x: BatchUpdateException => // ignore
+      case t: Throwable => throw t
+    }
+    insertRows3(20, serverHostPort, tableName, jdbcUser3, jdbcUser3)
+
+    try {
+      insertRows4(20, serverHostPort, jdbcUser3 + "." + tableName, jdbcUser4, jdbcUser4)
+      assert(false) // fail
+    } catch {
+      case x: SQLException if x.getSQLState.equals("42500") => // ignore
+      case t: Throwable => throw t
+    }
+    insertRows4(20, serverHostPort, tableName, jdbcUser3, jdbcUser3)
+
+    // (1 to 5).foreach(d => query())
+    try {
+      query2(serverHostPort, jdbcUser3 + "." + tableName, jdbcUser4, jdbcUser4)
+      assert(false) // fail
+    } catch {
+      case x: SQLException if x.getSQLState.equals("42502") => // ignore
+      case t: Throwable => throw t
+    }
+    query2(serverHostPort, tableName, jdbcUser3, jdbcUser3)
+
+    try {
+      dropTable2(serverHostPort, jdbcUser3 + "." + tableName, jdbcUser4, jdbcUser4)
+      assert(false) // fail
+    } catch {
+      case x: SQLException if x.getSQLState.equals("42502") => // ignore
+      case t: Throwable => throw t
+    }
+    dropTable2(serverHostPort, tableName, jdbcUser3, jdbcUser3)
+    rowTableRoutingCompleted.incrementAndGet()
+  }
+
+  def testConcurrency(): Unit = {
+    val colThread = new Thread(new Runnable {
+      def run() {
+        (1 to 5) foreach (i => {
+          // scalastyle:off println
+          println(s"columnTableRouting started $i")
+          // scalastyle:on println
+          val completed = columnTableRouting()
+          // scalastyle:off println
+          println(s"columnTableRouting finished $i $completed")
+          // scalastyle:on println
+        })
+      }
+    })
+    val rowThread = new Thread(new Runnable {
+      def run() {
+        (1 to 5) foreach (i => {
+          // scalastyle:off println
+          println(s"rowTableRouting started $i")
+          // scalastyle:on println
+          val completed = rowTableRouting()
+          // scalastyle:off println
+          println(s"rowTableRouting finished $i $completed")
+          // scalastyle:on println
+        })
+      }
+    })
+    colThread.start()
```

Comment:
Just 2 threads in parallel. One col and one row. It would be good to increase to 5 each at least.